File: harness/forge_harness/worktree_manager.py
"""
Git Worktree Manager
====================

Manages git worktrees for parallel agent execution.
Enables "1 human, multiple agents" by isolating agent workspaces.
"""

import logging
import subprocess
from dataclasses import dataclass
from pathlib import Path

logger = logging.getLogger(__name__)


@dataclass
class Worktree:
    path: Path
    branch: str
    head_hash: str


class WorktreeManager:
    """Manages git worktrees for a repository."""

    def __init__(self, repo_root: Path):
        self.repo_root = repo_root

    def list_worktrees(self) -> list[Worktree]:
        """List all worktrees."""
        try:
            result = subprocess.run(
                ["git", "worktree", "list", "--porcelain"],
                cwd=self.repo_root,
                capture_output=True,
                text=True,
                check=True,
            )

            worktrees = []
            current_worktree = {}

            for line in result.stdout.splitlines():
                if line.startswith("worktree "):
                    if current_worktree:
                        worktrees.append(self._parse_worktree(current_worktree))
                        current_worktree = {}
                    current_worktree["path"] = line.split(" ", 1)[1]
                elif line.startswith("branch "):
                    current_worktree["branch"] = line.split(" ", 1)[1].replace("refs/heads/", "")
                elif line.startswith("HEAD "):
                    current_worktree["head"] = line.split(" ", 1)[1]

            if current_worktree:
                worktrees.append(self._parse_worktree(current_worktree))

            return worktrees
        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to list worktrees: {e}")
            return []

    def _parse_worktree(self, data: dict) -> Worktree:
        return Worktree(
            path=Path(data.get("path", "")),
            branch=data.get("branch", "detached"),
            head_hash=data.get("head", ""),
        )

    def create_worktree(self, branch: str, path: Path | None = None) -> Path:
        """Create a new worktree for a branch.

        Args:
            branch: Branch name to check out
            path: Optional path for worktree. Defaults to ../{repo_name}-{branch}

        Returns:
            Path to the created worktree
        """
        if path is None:
            # Default convention: sibling directory
            repo_name = self.repo_root.name
            safe_branch = branch.replace("/", "-")
            path = self.repo_root.parent / f"{repo_name}-{safe_branch}"

        if path.exists():
            logger.info(f"Worktree path {path} already exists")
            return path

        try:
            # Create branch if it doesn't exist
            # Check if branch exists
            branch_exists = (
                subprocess.run(
                    ["git", "show-ref", "--verify", f"refs/heads/{branch}"],
                    cwd=self.repo_root,
                    capture_output=True,
                ).returncode
                == 0
            )

            cmd = ["git", "worktree", "add"]
            if not branch_exists:
                cmd.append("-b")
            cmd.extend([str(path), branch])

            subprocess.run(cmd, cwd=self.repo_root, check=True, capture_output=True)
            logger.info(f"Created worktree for {branch} at {path}")
            return path
        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to create worktree: {e.stderr}")
            raise RuntimeError(f"Failed to create worktree: {e.stderr}")

    def remove_worktree(self, path: Path, force: bool = False):
        """Remove a worktree."""
        try:
            cmd = ["git", "worktree", "remove", str(path)]
            if force:
                cmd.append("--force")

            subprocess.run(cmd, cwd=self.repo_root, check=True, capture_output=True)
        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to remove worktree: {e}")
            raise

    def prune(self):
        """Prune stale worktree information."""
        subprocess.run(["git", "worktree", "prune"], cwd=self.repo_root, check=True)

File: cmd/forged/worktree_manager.go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// WorktreeInfo represents a single git worktree entry.
type WorktreeInfo struct {
	Path   string
	Branch string
	Hash   string
}

// WorktreeManager manages git worktrees for isolated task execution, as
// described in ADR-024 (Orchestration-Level Worktree Isolation).
type WorktreeManager struct {
	repoRoot    string
	worktreeDir string // Typically .forge/worktrees/
	db          *sql.DB
}

// NewWorktreeManager creates a new WorktreeManager for the given repository
// root. If worktreeDir is empty, it defaults to ".forge/worktrees" under the
// repo root.
func NewWorktreeManager(repoRoot, worktreeDir string, db *sql.DB) *WorktreeManager {
	if repoRoot == "" {
		repoRoot, _ = os.Getwd()
	}
	if worktreeDir == "" {
		worktreeDir = filepath.Join(repoRoot, ".forge", "worktrees")
	}
	return &WorktreeManager{
		repoRoot:    repoRoot,
		worktreeDir: worktreeDir,
		db:          db,
	}
}

// CreateWorktree creates an isolated git worktree for the given task ID.
// It returns the filesystem path to the worktree.
//
// git worktree add .forge/worktrees/{taskID} -b feature/{taskID}
func (m *WorktreeManager) CreateWorktree(ctx context.Context, taskID string) (string, error) {
	if strings.TrimSpace(taskID) == "" {
		return "", fmt.Errorf("taskID is required")
	}

	if err := os.MkdirAll(m.worktreeDir, 0o755); err != nil {
		return "", fmt.Errorf("create worktree dir: %w", err)
	}

	path := filepath.Join(m.worktreeDir, taskID)
	branch := fmt.Sprintf("feature/%s", taskID)

	// If the directory already exists, assume the worktree is present.
	if fi, err := os.Stat(path); err == nil && fi.IsDir() {
		if resolved, err := filepath.EvalSymlinks(path); err == nil {
			return resolved, nil
		}
		return path, nil
	}

	args := []string{"-C", m.repoRoot, "worktree", "add", path, "-b", branch}
	cmd := exec.CommandContext(ctx, "git", args...)
	output, err := cmd.CombinedOutput()
	if err != nil {
		outStr := string(output)
		// Handle already-exists / already-registered scenarios gracefully.
		if strings.Contains(outStr, "already exists") || strings.Contains(outStr, "already registered") {
			if resolved, err := filepath.EvalSymlinks(path); err == nil {
				return resolved, nil
			}
			return path, nil
		}
		return "", fmt.Errorf("git worktree add failed: %v (output: %s)", err, outStr)
	}

	if resolved, err := filepath.EvalSymlinks(path); err == nil {
		return resolved, nil
	}
	return path, nil
}

// RemoveWorktree removes the worktree for the given task ID and attempts to
// delete the corresponding branch. It is tolerant of missing worktrees or
// branches and performs best-effort cleanup.
func (m *WorktreeManager) RemoveWorktree(ctx context.Context, taskID string) error {
	if strings.TrimSpace(taskID) == "" {
		return fmt.Errorf("taskID is required")
	}

	path := filepath.Join(m.worktreeDir, taskID)
	branch := fmt.Sprintf("feature/%s", taskID)

	// Remove the worktree (force to handle dirty worktrees).
	args := []string{"-C", m.repoRoot, "worktree", "remove", "--force", path}
	cmd := exec.CommandContext(ctx, "git", args...)
	if output, err := cmd.CombinedOutput(); err != nil {
		// If git refuses because it does not recognise the worktree, fall back
		// to direct filesystem removal.
		if _, statErr := os.Stat(path); statErr == nil {
			_ = os.RemoveAll(path)
		}
		_ = output // log hook can be added later; ignore for now
	} else {
		_ = cmd // keep for parity; no-op
	}

	// Best-effort branch deletion; ignore errors if branch does not exist or
	// is not fully merged.
	delArgs := []string{"-C", m.repoRoot, "branch", "-D", branch}
	delCmd := exec.CommandContext(ctx, "git", delArgs...)
	_, _ = delCmd.CombinedOutput()

	// Ensure directory is gone even if git did not clean it up.
	if _, err := os.Stat(path); err == nil {
		if rmErr := os.RemoveAll(path); rmErr != nil {
			return fmt.Errorf("remove worktree dir: %w", rmErr)
		}
	}

	return nil
}

// ListWorktrees returns the active git worktrees by parsing:
//   git worktree list --porcelain
func (m *WorktreeManager) ListWorktrees(ctx context.Context) ([]WorktreeInfo, error) {
	args := []string{"-C", m.repoRoot, "worktree", "list", "--porcelain"}
	cmd := exec.CommandContext(ctx, "git", args...)
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("git worktree list failed: %w", err)
	}

	lines := strings.Split(string(output), "\n")
	var (
		result []WorktreeInfo
		cur    *WorktreeInfo
	)

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		switch {
		case strings.HasPrefix(line, "worktree "):
			// Flush previous entry.
			if cur != nil {
				result = append(result, *cur)
			}
			rawPath := strings.TrimSpace(strings.TrimPrefix(line, "worktree "))
			// Resolve symlinks so callers can compare with os.* paths (macOS: /var → /private/var).
			if resolved, err := filepath.EvalSymlinks(rawPath); err == nil {
				rawPath = resolved
			}
			cur = &WorktreeInfo{
				Path: rawPath,
			}
		case strings.HasPrefix(line, "HEAD "):
			if cur != nil {
				cur.Hash = strings.TrimSpace(strings.TrimPrefix(line, "HEAD "))
			}
		case strings.HasPrefix(line, "branch "):
			if cur != nil {
				ref := strings.TrimSpace(strings.TrimPrefix(line, "branch "))
				cur.Branch = strings.TrimPrefix(ref, "refs/heads/")
			}
		default:
			// ignore other porcelain lines (locked, prunable, etc.)
		}
	}

	if cur != nil {
		result = append(result, *cur)
	}

	return result, nil
}

// RecordWorktree inserts a worktree row without touching the filesystem.
// Used by the daemon to track worktrees created by agent scripts.
func (m *WorktreeManager) RecordWorktree(ctx context.Context, taskID, path, branch, agentID string) error {
	_, err := m.db.ExecContext(ctx,
		`INSERT OR REPLACE INTO worktrees (id, path, branch, agent_id, status, created_at)
		 VALUES (?, ?, ?, ?, 'active', datetime('now'))`,
		taskID, path, branch, agentID,
	)
	return err
}

// IsWorktreeDirty checks if the worktree has uncommitted changes (including untracked files).
func (m *WorktreeManager) IsWorktreeDirty(ctx context.Context, taskID string) (bool, error) {
	path := filepath.Join(m.worktreeDir, taskID)
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return false, nil
	}

	// git status --porcelain shows all changes (staged, unstaged, untracked)
	args := []string{"-C", path, "status", "--porcelain"}
	cmd := exec.CommandContext(ctx, "git", args...)
	output, err := cmd.Output()
	if err != nil {
		// If it's not a git repo or other error, we can't reliably say it's clean.
		return true, fmt.Errorf("git status failed: %w", err)
	}

	return len(strings.TrimSpace(string(output))) > 0, nil
}

// PruneStaleWorktrees removes worktrees for completed/failed tasks. It is
// intended to be called periodically by patrols. If db is nil, it performs
// a pure filesystem-based cleanup of orphaned directories.
// It preserves "dirty" worktrees even if they are stale to prevent data loss.
func (m *WorktreeManager) PruneStaleWorktrees(ctx context.Context) error {
	entries, err := os.ReadDir(m.worktreeDir)
	if err != nil {
		// If the directory does not exist yet, there is nothing to prune.
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("read worktree dir: %w", err)
	}

	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		taskID := e.Name()

		shouldRemove := false
		if m.db == nil {
			// Without DB, treat all entries as candidates for pruning only if
			// their directories are empty.
			path := filepath.Join(m.worktreeDir, taskID)
			if dirEntries, _ := os.ReadDir(path); len(dirEntries) == 0 {
				shouldRemove = true
			}
		} else {
			var status string
			err := m.db.QueryRowContext(ctx,
				`SELECT status FROM tasks WHERE id = ?`,
				taskID,
			).Scan(&status)
			if err == sql.ErrNoRows {
				// No corresponding task; safe to prune.
				shouldRemove = true
			} else if err != nil {
				// On DB error, skip this worktree and continue.
				continue
			} else {
				if status == string(TaskStatusCompleted) || status == string(TaskStatusFailed) {
					shouldRemove = true
				}
			}
		}

		if shouldRemove {
			// Verification: don't prune dirty worktrees (Proposal #3 in S157 retro).
			dirty, err := m.IsWorktreeDirty(ctx, taskID)
			if err == nil && dirty {
				// Log or skip? Proposal says "Warn and preserve".
				// For now, we just preserve by skipping.
				continue
			}
			_ = m.RemoveWorktree(ctx, taskID)
		}
	}

	return nil
}

